### PR TITLE
Enable Gemma 4 E2B / E4B inference via vLLM RPA

### PR DESCRIPTION
**Pull request by gagika**
## Description

Enable Gemma 4 E2B / E4B inference via vLLM adapter by adding KV-shared layers wiring and a `system_prompt` flag to `vllm_decode` (required by the E-family `-it` checkpoints), and documents a verified inference recipe in `Run_Gemma4.md`.

## Key changes

* `gemma4_small.py` — decoder layer returns the kernel-updated `kv_cache` (was dropped).
* `decoders.py` — KV-shared layers redirect to the donor's `kv_caches` slot via a layer→slot map; cache is written back per layer.
* `attentions.py` — sliding-window only on `LOCAL_SLIDING` layers; KV-shared layers no longer overwrite the donor's cache (`update_kv_cache=False`).
* `vllm_decode.py` / `types.py` — new `system_prompt` config knob, prepended as a system message when `use_chat_template=True`. Adapter registration + env flags moved from import time into main() (import side effects leaked into unrelated tests).
* `Run_Gemma4.md` — E2B / E4B recipe: system prompt, model-card sampling, full `eos_token_id` `[1, 106, 50]` stop-token set.

# Tests

On v5p-8, e2b + e4b: cross-checked top-1 logits at greedy vs the native checkpoint — bit-identical. With the documented recipe, models generate coherent output and stop cleanly. CLI and Python API verified.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/decoders.py [hunk 1]
```diff
@@ -1397,6 +1401,18 @@ def _apply_gemma4_small_layers(
...
 
+    # tpu-inference allocates one `kv_caches` slot per non-shared layer;
+    # KV-shared layers reuse the donor's slot. Map decoder layer index -> slot.
+    cache_index_of: dict[int, int] = {}
+    next_slot = 0
+    for lyr in range(cfg.num_decoder_layers):
+      donor_idx = gemma4_small.kv_donor_layer_idx(lyr, layer_types, num_kv_shared)
+      if donor_idx is not None:
+        cache_index_of[lyr] = cache_index_of[donor_idx]
+      else:
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/layers/decoders.py]: The map is built only at JIT trace time (a handful of traces per process), not per decode step.

**Review by khatwanimohit — APPROVED** [review]
LGTM

**Review by NuojCheng — APPROVED** [review]
